config/derive/src/proxy.rs

```diff
@@ -0,0 +1,214 @@
+
+    let set_field_methods = builder_resolver.set_field_methods;
+    let field_names = builder_resolver.field_names;
+
+    let impl_parent_default = if !builder_resolver.has_required_field {
+        let ident = field_names
+            .iter()
+            .map(|field_value| format_ident!("DEFAULT_{}", field_value.to_string().to_uppercase()));
+
+        quote! {
+            impl Default for #parent_ident {
+                fn default() -> Self {
+                    Configuration::new()
+                }
+            }
+
+            impl #parent_ident {
+                pub const fn new() -> Self {
+                    Self { #( #field_names: Self::#ident ),* }
+                }
+            }
+        }
+    } else {
+        quote!()
+    };
+
+    let builder_default_field_values: Vec<_> = builder_resolver
+        .default_field_values
+        .into_iter()
+        .enumerate()
+        .map(|(i, field_value)| {
+            let field_name = &field_names[i];
+            field_value.map_or_else(
+                || quote! {#field_name: None},
+                |field_value| {
+                    let ident = format_ident!("DEFAULT_{}", field_value.to_string().to_uppercase());
+
+                    quote! {
+                        #field_name: Some(#parent_ident::#ident)
+                    }
+                },
+            )
+        })
+        .collect();
+
+    quote! {
+        #impl_parent_default
+
+        /// [`Configuration`] builder
+        #[derive(
+            Debug, Clone,
+            serde::Deserialize, serde::Serialize,
+            iroha_config_derive::Documented,
+        )]
+        #[serde(default)]
+        #builder_struct
+
+        impl Default for #proxy_struct_name {
+            fn default() -> Self {
+                Self::new()
+            }
+        }
+
+        impl #proxy_struct_name {
+            pub const fn new() -> Self {
+                Self { #( #builder_default_field_values ),* }
+            }
+
+            #( #set_field_methods )*
+
+            /// Override [`Self`] with values from another [`Self`]
+            pub fn override_with(&mut self, other: Self) -> &mut Self { #(
+                if let Some(#field_names) = other.#field_names {
+                    // TODO: Delegate to setter method instead
+                    self.#field_names = Some(#field_names);
+                }; )*
+
+                self
+            }
+
+            /// Build [`Configuration`]
+            ///
+            /// # Errors
+            ///
+            /// - if missing required fields
+            pub fn build(self) -> Result<#parent_ident, ::iroha_config::Error> {
+                let config = Self::default().override_with(self);
+
+                Ok(#parent_ident { #(
+                    #field_names: config.#field_names.ok_or(
+                        ::iroha_config::Error::MissingField(stringify!(#field_names))
+                    )? ),*
+                })
+            }
```

Comment:
It seems that the final `Configuration` should not have any `None`s, otherwise you will get a missing field error. Is it the case? Isn't it possible that the final configuration could have an empty field that is not defined neither by user and by defaults?

Reply:
These questions feel like they need to be phrased as tests. 

One thing we found was useful was pair programming. You could contribute those directly to @mversic 's fork and expedite the invariant discussion.